pvlib/tools.py

```diff
@@ -345,15 +349,18 @@ def _golden_sect_DataFrame(params, lower, upper, func, atol=1e-8):
         err = abs(df['V2'] - df['V1'])
 
         # works with single value because err is np.float64
-        converged = (err < atol).all()
+        converged = (err[~np.isnan(err)] < atol).all()
         # err will be less than atol before iterations hit the limit
         # but just to be safe
         iterations += 1
 
     if iterations > iterlimit:
-        raise Exception("iterations exceeded maximum")  # pragma: no cover
+        raise Exception("Iterations exceeded maximum. Check that func",
+                        " is not NaN in (lower, upper)")
 
-    return func(df, 'V1'), df['V1']
+    func_result = func(df, 'V1')
```

Comment:
This PR gets #1394 working again for me locally, so +1 for that.  However I think this will still fail with a KeyError if either of `upper` or `lower` is all nans: in that case `iterlimit` will still be nan, so the while loop will not execute at all, causing `df['V1']` to not exist when this line executes.  I think the all-nan case is worth supporting:  a ModelChain simulation of only nighttime is not particularly useful, but shouldn't throw an error IMHO.